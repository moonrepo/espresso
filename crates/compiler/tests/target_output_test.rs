mod utils;

use espresso_common::EsTarget;
use espresso_compiler::Compiler;
use espresso_package::Package;
use espresso_store::Store;
use starbase_sandbox::{assert_snapshot, create_sandbox};
use std::sync::Arc;
use utils::*;

macro_rules! test_target {
    ($method:ident, $target:expr) => {
        mod $method {
            use super::*;

            #[tokio::test]
            async fn transforms_modules() {
                let sandbox = create_sandbox("syntax");
                let package = Package::new(sandbox.path()).unwrap();
                let compiler = create_compiler(sandbox.path(), &package);

                let out_dir = compiler.compile($target).await.unwrap();

                assert_snapshot!(read_file(out_dir.join("index.mjs")));
            }

            #[tokio::test]
            async fn creates_tsconfig() {
                let sandbox = create_sandbox("syntax");
                let package = Package::new(sandbox.path()).unwrap();
                let compiler = create_compiler(sandbox.path(), &package);

                let out_dir = compiler.compile($target).await.unwrap();
                let tsconfig = format!("tsconfig.{}.json", $target.to_string());

                assert_snapshot!(read_file(out_dir.join("..").join(tsconfig)));
            }
        }
    };
}

mod target_output {
    use super::*;

    test_target!(es2015, EsTarget::Es2015);
    test_target!(es2016, EsTarget::Es2016);
    test_target!(es2017, EsTarget::Es2017);
    test_target!(es2018, EsTarget::Es2018);
    test_target!(es2019, EsTarget::Es2019);
    test_target!(es2020, EsTarget::Es2020);
    test_target!(es2021, EsTarget::Es2021);
    test_target!(es2022, EsTarget::Es2022);

    #[tokio::test]
    async fn supports_legacy_decorators() {
        let sandbox = create_sandbox("syntax-legacy-decorators");
        let package = Package::new(sandbox.path()).unwrap();
        let compiler = Compiler::new(
            &package,
            Arc::new(Store::load_from(sandbox.path()).unwrap()),
        )
        .unwrap();
        let out_dir = compiler.compile(EsTarget::Es2018).await.unwrap();

        assert_snapshot!(read_file(out_dir.join("index.mjs")));
        assert_snapshot!(read_file(out_dir.join("other.mjs")));
    }
}
