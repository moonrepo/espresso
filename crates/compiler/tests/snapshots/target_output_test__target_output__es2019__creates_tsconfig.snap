---
source: crates/compiler/tests/target_output_test.rs
expression: "read_file(out_dir.join(\"..\").join(tsconfig))"
---
{
  "compilerOptions": {
    "allowArbitraryExtensions": true,
    "allowImportingTsExtensions": true,
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "declaration": true,
    "declarationMap": false,
    "emitDeclarationOnly": true,
    "esModuleInterop": true,
    "experimentalDecorators": false,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "lib": [
      "dom",
      "es2019",
    ],
    "module": "es2015",
    "moduleResolution": "nodenext",
    "noEmitOnError": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "outDir": "./es2019",
    "paths": {},
    "pretty": true,
    "removeComments": false,
    "resolvePackageJsonExports": false,
    "resolvePackageJsonImports": false,
    "resolveJsonModule": true,
    "rootDir": "../src",
    "skipLibCheck": true,
    "sourceMap": false,
    "strict": true,
    "target": "es2019",
    "verbatimModuleSyntax": true
  },
  "include": ["../src/**/*"]
}

